deeplake/core/dataset/deeplake_query_dataset.py

```diff
@@ -0,0 +1,375 @@
+import posixpath
+from typing import Any, Callable, Dict, List, Optional, Sequence, Tuple, Union
+from functools import partial
+from deeplake.constants import SHOW_ITERATION_WARNING
+
+from time import time
+import math
+
+from deeplake.util.iteration_warning import (
+    check_if_iteration,
+)
+from deeplake.api.info import load_info
+from deeplake.client.log import logger
+from deeplake.client.utils import get_user_name
+from deeplake.constants import (
+    SAMPLE_INFO_TENSOR_MAX_CHUNK_SIZE,
+)
+from deeplake.core.index import Index
+from deeplake.core.tensor import Tensor
+from deeplake.util.bugout_reporter import deeplake_reporter
+from deeplake.util.exceptions import (
+    InvalidKeyTypeError,
+    InvalidOperationError,
+    MemoryDatasetCanNotBePickledError,
+    TensorDoesNotExistError,
+)
+from deeplake.util.scheduling import calculate_absolute_lengths
+from deeplake.core.dataset import Dataset
+
+import warnings
+
+from deeplake.core.dataset.deeplake_query_tensor import DeepLakeQueryTensor
+
+
+class DeepLakeQueryDataset(Dataset):
+    def __init__(self, deeplake_ds, indra_ds, group_index=None, enabled_tensors=None):
+        self.deeplake_ds = deeplake_ds
+        self.indra_ds = indra_ds
+        self.group_index = group_index or deeplake_ds.group_index
+        self.enabled_tensors = enabled_tensors or deeplake_ds.enabled_tensors
+        self.set_deeplake_dataset_variables()
+
+    def set_deeplake_dataset_variables(self):
+        self.path = self.deeplake_ds.path
+        if self.path.startswith("mem://"):
+            raise MemoryDatasetCanNotBePickledError
+
+        keys = [
+            "base_storage",
+            "_read_only",
+            "public",
+            "storage",
+            "_token",
+            "verbose",
+            "version_state",
+            "org_id",
+            "ds_name",
+            "_is_filtered_view",
+            "_view_id",
+            "_view_use_parent_commit",
+            "_parent_dataset",
+            "_pad_tensors",
+            "_locking_enabled",
+            "is_iteration",
+            "_view_base",
+            "link_creds",
+            "_locked_out",
+            "_indexing_history",
+            "_vc_info_updated",
+        ]
+
+        for k in keys:
+            setattr(self, k, getattr(self.deeplake_ds, k))
```

Comment:
This wouldnt reflect the changes made to `deeplake_ds` later on. Better to override `getattr` and `setattr`.